### PR TITLE
Support thrift 0.19 and 0.20

### PR DESCRIPTION
**Pull request by jrhee17**
Motivation:

Support for thrift 0.19/0.20 has been requested internally.

Notably, there has been two changes upstream.
- Jakarta EE, Apache HttpClient 5 is now used [ref](https://github.com/apache/thrift/pull/2746)
- The "true" type is now used when generating Metadata [ref](https://github.com/apache/thrift/pull/2765/files)

For the first change, test files have been modified so that HttpClient/Jakarta related classes can be overwritten.

The second change is interesting because generated messages for `typdef` enums are now correctly set as an `EnumMetadata`. As a result of this, using `TTextProtocol` with `useNamedEnums=true` now yields the correct output (`typedef` enums serialized as the enum name). This is consistent with the other enum serialization results as well.

One downside of this is older version messages (generated with thrift<0.19) cannot deserialize a typedef enum field by enum name. This is because a `FieldValueMetaData` is used, and it is not possible to know which enum the field is linked to unless we parse the thrift file manually.

Modifications:

- Jakarta EE, Apache HttpClient 5 related dependencies are added
  - Extracted `ServletTestUtils` so that `ThriftOverHttpClientTServletIntegrationTest` can be shared across modules.
  - Extracted `ApacheClientUtils` so that `ThriftDocServiceTest`, `ThriftOverHttp1Test` can be shared across modules.
  - Modified to use JDK over Apache's `Base64` library
- Generated messages for `typdef` enums are now correctly set, so `TTextProtocolTest#tTextNamedEnumProtocolReadWriteTest` has been modified to check if the generated `typedef` value is a enum ordinal or name.

Result:

- https://github.com/line/armeria/issues/5243

### CODE REVIEW

File: thrift/thrift0.12/src/test/thrift/TTextProtocolTest.thrift [hunk 1]
```diff
@@ -104,7 +104,7 @@ struct TTextProtocolTestMsg {
 
   22: required list<TestUnion> x;
 
-  23: Moji y;  // Does not support string values.
+  23: Moji y;  // Does not support string values for thrift version < 0.19.
```
Comment (minwoox) [inline thrift/thrift0.12/src/test/thrift/TTextProtocolTest.thrift:107]: Is there any test case for this from 0.19?

**Review by minwoox — APPROVED** [review]
Thanks a lot, @jrhee17! 👍 👍 👍

File: thrift/thrift0.13/src/test/java/com/linecorp/armeria/common/thrift/text/TTextProtocolTest.java [hunk 2]
```diff
@@ -139,9 +141,14 @@ void tTextNamedEnumProtocolReadWriteTest() throws Exception {
...
         final byte[] bytes = baos.toByteArray();
 
+        // ignore y since it can be serialized to either 1 (<0.19) or ALPHA (>=0.19)
         assertThatJson(CR_PATTERN.matcher(baos.toString()).replaceAll(""))
                 .when(IGNORING_ARRAY_ORDER)
+                .whenIgnoringPaths("y")
                 .isEqualTo(namedEnumSerialized);
+        assertThatJson(CR_PATTERN.matcher(baos.toString()).replaceAll(""))
+                .node("y")
+                .matches(anyOf(hasToString("1"), hasToString("ALPHA")));
```
Comment (ikhoon) [inline thrift/thrift0.13/src/test/java/com/linecorp/armeria/common/thrift/text/TTextProtocolTest.java:151]: Suggestion) Should we match the exact value depending on `:thrift` module versions?
We could read the version from `/META-INF/com.linecorp.armeria.versions.properties` in runtime.

**Review by ikhoon — APPROVED** [review]
Thanks, @jrhee17! 🙇‍♂️

**Comment by KarboniteKream** [comment]
Thanks, I'll test it today. You just want to make sure that `TTEXT` still works for our CMS APIs?

**Comment by jrhee17** [comment]
Serialized messages using `TTextProtocol` + `typdef Enum` + `useNamedEnums=true`  using thrift compiler >=0.19 can't be read by remotes using older messages (<0.19).

e.g. If a client with (`TTextProtocol` + `typdef Enum` + `useNamedEnums=true` + `>=0.19` generated message) sends a request to a server (<0.19) then the server won't be able to parse the message.

This may or may not be a problem depending on your use-case.

**Comment by KarboniteKream** [comment]
In this specific case, the client and the server are using the same Thrift version, so this should not be causing issues.

File: thrift/thrift0.13/src/test/java/com/linecorp/armeria/common/thrift/text/TTextProtocolTest.java (hunk 2 above)
Comment (jrhee17) [reply thrift/thrift0.13/src/test/java/com/linecorp/armeria/common/thrift/text/TTextProtocolTest.java]: Updated. While doing so, I found that the thrift artifact versions weren't fetched properly due to having a `.`, so `lastIndexOf` is used in `Versions`

File: thrift/thrift0.12/src/test/thrift/TTextProtocolTest.thrift (hunk 1 above)
Comment (jrhee17) [reply thrift/thrift0.12/src/test/thrift/TTextProtocolTest.thrift:107]: The test at https://github.com/line/armeria/pull/5822#discussion_r1689777662 has been modified to better reflect this change

File: thrift/thrift0.13/src/test/java/com/linecorp/armeria/internal/server/thrift/ThriftDocServiceTest.java [hunk 3]
```diff
@@ -257,12 +256,10 @@ public void excludeAllServices() throws IOException {
...
-        try (CloseableHttpClient hc = HttpClients.createMinimal()) {
-            final HttpPost req = new HttpPost(server.httpUri() + "/docs/specification.json");
-
-            try (CloseableHttpResponse res = hc.execute(req)) {
-                assertThat(res.getStatusLine().toString()).isEqualTo("HTTP/1.1 405 Method Not Allowed");
-            }
-        }
+        final Map.Entry<String, String> res =
+                ApacheClientUtils.makeApacheHttpRequest(server.httpUri() + "/docs/specification.json",
+                                                        HttpMethod.POST);
```
[the remark this answers is not in the document]
Comment (jrhee17) [reply thrift/thrift0.13/src/test/java/com/linecorp/armeria/internal/server/thrift/ThriftDocServiceTest.java]: I've modified the Apache http client usages to use Armeria's `WebClient`. Additionally, I've migrated the junit4 test to junit5 🙇

[the remark this answers is not in the document]
Comment (ikhoon) [reply thrift/thrift0.13/src/test/java/com/linecorp/armeria/internal/server/thrift/ThriftDocServiceTest.java]: Thank you very much. ❤️ I didn't mean to fix this though.

**Pull request merged**